src/main/java/decodes/cwms/CwmsTimeSeriesDAO.java

```diff
@@ -1285,4 +1298,302 @@ public void modifyTSID(TimeSeriesIdentifier tsid)
+				return dataCollection;
+			}
+			else
+				minRecNum = rs.getInt(1);
+			if (rs.wasNull())
+			{
+				debug1("No new data for appId=" + applicationId);
+				minRecNum = -1;
+				((TimeSeriesDb)db).reclaimTasklistSpace(this);
+				return dataCollection;
+			}
+		}
+		catch(SQLException ex)
+		{
+			warning("getNewDataSince: " + ex);
+			return dataCollection;
+		}
+
+		ArrayList<TasklistRec> tasklistRecs = new ArrayList<TasklistRec>();
+		ArrayList<Integer> badRecs = new ArrayList<Integer>();
+		try
+		{
+			getTaskListStmt.setInt(1, minRecNum);
+
+			int maxRecNum = minRecNum + 10000;
+			if (maxRecNum < minRecNum)
+			{
+				// The 32-bit integer wrapped around. Set to max possible int.
+				maxRecNum = Integer.MAX_VALUE;
+			}
+
+			debug3("Executing '" + getTaskListStmtQuery + "'");
+			ResultSet rs = getTaskListStmt.executeQuery();
+			while (rs.next())
+			{
+				// Extract the info needed from the result set row.
+				int recordNum = rs.getInt(1);
+				DbKey sdi = DbKey.createDbKey(rs, 2);
+				double value = rs.getDouble(3);
+				boolean valueWasNull = rs.wasNull();
+				Date timeStamp = db.getFullDate(rs, 4);
+				String df = rs.getString(5);
+				char c = df.toLowerCase().charAt(0);
+				boolean deleted = false;
+				if (c == 'u')
+				{
+					// msg handler will send this when he gets
+					// TsCodeChanged. It tells me to update my cache.
+					DbObjectCache<TimeSeriesIdentifier> cache = getCache();
+					synchronized(cache)
+					{
+						TimeSeriesIdentifier tsid = cache.getByKey(sdi);
+						if (tsid != null)
+						{
+							DbKey newCode = DbKey.createDbKey(rs, 9);
+							cache.remove(sdi);
+							tsid.setKey(newCode);
+							cache.put(tsid);
+							continue;
+						}
+					}
+				}
+				else
+					deleted = TextUtil.str2boolean(df);
+
+				String unitsAbbr = rs.getString(6);
+				Date versionDate = db.getFullDate(rs, 7);
+				BigDecimal qc = rs.getBigDecimal(8);
+				long qualityCode = qc == null ? 0 : qc.longValue();
+
+				TasklistRec rec = new TasklistRec(recordNum, sdi, value,
+					valueWasNull, timeStamp, deleted,
+					unitsAbbr, versionDate, qualityCode);
+				tasklistRecs.add(rec);
+			}
+
+			RecordRangeHandle rrhandle = new RecordRangeHandle(applicationId);
+
+			// Process the real-time records collected above.
+			for(TasklistRec rec : tasklistRecs)
+				processTasklistEntry(rec, dataCollection, rrhandle, badRecs, applicationId);
+
+			dataCollection.setTasklistHandle(rrhandle);
+
+			// Delete the bad tasklist recs, 250 at a time.
+			if (badRecs.size() > 0)
+				Logger.instance().debug1("getNewDataSince deleting " + badRecs.size()
+					+ " bad tasklist records.");
+			while (badRecs.size() > 0)
+			{
+				StringBuilder inList = new StringBuilder();
+				int n = badRecs.size();
+				int x=0;
+				for(; x<250 && x<n; x++)
+				{
+					if (x > 0)
+						inList.append(", ");
+					inList.append(badRecs.get(x).toString());
+				}
+				String q = "delete from CP_COMP_TASKLIST "
```

Comment:
This should be the change I had made earlier. Currently, this create a new query in oracle's (presumably Postgres works similarly) shared memory space each time it's called; and it's different every time. Lead to a lot of performance issues as it's reevaluating OTHER query plans way more often.

Reply:
For the curious, i learned last week while researching for another project that PostgreSQL doesn't actually have such a shared query pool; only kept per session.